File: src/unix/ibus/config_util_test.cc
// Copyright 2010-2011, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

// Test for config_util

#include "unix/ibus/config_util.h"

#include "testing/base/public/gunit.h"
#include "session/config.pb.h"

TEST(ConfigUtilTest, IntValue) {
  mozc::config::Config config;
  EXPECT_EQ("", config.DebugString());

#if IBUS_CHECK_VERSION(1, 3, 99)
  GVariant *value = g_variant_new_int32(42);
#else
  g_type_init();
  GValue actual_value = {0};
  GValue *value = &actual_value;
  g_value_init(value, G_TYPE_INT);
  g_value_set_int(value, 42);
#endif
  mozc::ibus::ConfigUtil::SetFieldForName(
      "config_version", value, &config);

  EXPECT_EQ("config_version: 42\n",
            config.DebugString());
#if IBUS_CHECK_VERSION(1, 3, 99)
  g_variant_unref(value);
#else
  g_value_unset(value);
#endif
}

TEST(ConfigUtilTest, EnumValue) {
  mozc::config::Config config;
  EXPECT_EQ("", config.DebugString());

#if IBUS_CHECK_VERSION(1, 3, 99)
  GVariant *value = g_variant_new_string("ROMAN");
#else
  g_type_init();
  GValue actual_value = {0};
  GValue *value = &actual_value;
  g_value_init(value, G_TYPE_STRING);
  g_value_set_static_string(value, "ROMAN");
#endif
  mozc::ibus::ConfigUtil::SetFieldForName(
      "preedit_method", value, &config);

  EXPECT_EQ("preedit_method: ROMAN\n",
            config.DebugString());
#if IBUS_CHECK_VERSION(1, 3, 99)
  g_variant_unref(value);
#else
  g_value_unset(value);
#endif
}

TEST(ConfigUtilTest, BooleanValue) {
  mozc::config::Config config;
  EXPECT_EQ("", config.DebugString());

#if IBUS_CHECK_VERSION(1, 3, 99)
  GVariant *value = g_variant_new_boolean(true);
#else
  g_type_init();
  GValue actual_value = {0};
  GValue *value = &actual_value;
  g_value_init(value, G_TYPE_BOOLEAN);
  g_value_set_boolean(value, true);
#endif
  mozc::ibus::ConfigUtil::SetFieldForName(
      "incognito_mode", value, &config);

  EXPECT_EQ("incognito_mode: true\n",
            config.DebugString());
#if IBUS_CHECK_VERSION(1, 3, 99)
  g_variant_unref(value);
#else
  g_value_unset(value);
#endif
}

TEST(ConfigUtilTest, NoSuchField) {
  mozc::config::Config config;
  EXPECT_EQ("", config.DebugString());

#if IBUS_CHECK_VERSION(1, 3, 99)
  GVariant *value = g_variant_new_int32(10);
#else
  g_type_init();
  GValue actual_value = {0};
  GValue *value = &actual_value;
  g_value_init(value, G_TYPE_INT);
  g_value_set_int(value, 10);
#endif
  mozc::ibus::ConfigUtil::SetFieldForName(
      "no_such_field", value, &config);

  // Does not affect the config value.
  EXPECT_EQ("", config.DebugString());
#if IBUS_CHECK_VERSION(1, 3, 99)
  g_variant_unref(value);
#else
  g_value_unset(value);
#endif
}

TEST(ConfigUtilTest, TypeMismatch) {
  mozc::config::Config config;
  EXPECT_EQ("", config.DebugString());

#if IBUS_CHECK_VERSION(1, 3, 99)
  GVariant *value = g_variant_new_int32(10);
#else
  g_type_init();
  GValue actual_value = {0};
  GValue *value = &actual_value;
  g_value_init(value, G_TYPE_INT);
  g_value_set_int(value, 10);
#endif

  // enum field
  mozc::ibus::ConfigUtil::SetFieldForName(
      "preedit_method", value, &config);

  // Does not affect the config value.
  EXPECT_EQ("", config.DebugString());
#if IBUS_CHECK_VERSION(1, 3, 99)
  g_variant_unref(value);
#else
  g_value_unset(value);
#endif
}

File: src/android/tests/src/com/google/android/inputmethod/japanese/ConfigUtilTest.java
// Copyright 2010-2013, Google Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are
// met:
//
//     * Redistributions of source code must retain the above copyright
// notice, this list of conditions and the following disclaimer.
//     * Redistributions in binary form must reproduce the above
// copyright notice, this list of conditions and the following disclaimer
// in the documentation and/or other materials provided with the
// distribution.
//     * Neither the name of Google Inc. nor the names of its
// contributors may be used to endorse or promote products derived from
// this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package org.mozc.android.inputmethod.japanese;

import org.mozc.android.inputmethod.japanese.preference.PreferenceUtil;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.Config;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.Config.FundamentalCharacterForm;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.Config.HistoryLearningLevel;
import org.mozc.android.inputmethod.japanese.protobuf.ProtoConfig.GeneralConfig;
import org.mozc.android.inputmethod.japanese.testing.MozcPreferenceUtil;
import org.mozc.android.inputmethod.japanese.testing.Parameter;

import android.content.SharedPreferences;
import android.test.InstrumentationTestCase;
import android.test.suitebuilder.annotation.SmallTest;

/**
 */
public class ConfigUtilTest extends InstrumentationTestCase {
  @SmallTest
  public void testToConfig_fundamentalCharacterFormFromBooleanKey() {
    class TestData extends Parameter {
      public final boolean halfSpace;
      public final Config expectedConfig;
      public TestData(boolean halfSpace, FundamentalCharacterForm expectedForm) {
        this.halfSpace = halfSpace;
        this.expectedConfig = Config.newBuilder().setSpaceCharacterForm(expectedForm).build();
      }
    }
    TestData[] testDataList = {
        new TestData(true, FundamentalCharacterForm.FUNDAMENTAL_HALF_WIDTH),
        new TestData(false, FundamentalCharacterForm.FUNDAMENTAL_INPUT_MODE),
    };

    SharedPreferences sharedPreferences = MozcPreferenceUtil.getSharedPreferences(
        getInstrumentation().getContext(), "TEST_FOR_FUNDAMENTAL_CHAR_FORM");

    for (TestData testData : testDataList) {
      sharedPreferences.edit()
          .clear()
          .putBoolean(PreferenceUtil.PREF_SPACE_CHARACTER_FORM_KEY, testData.halfSpace)
          .commit();

      assertEquals(testData.toString(),
                   testData.expectedConfig, ConfigUtil.toConfig(sharedPreferences));
    }
  }

  @SmallTest
  public void testToConfig_kanaModifierInsensitiveConversion() {
    SharedPreferences sharedPreferences = MozcPreferenceUtil.getSharedPreferences(
        getInstrumentation().getContext(), "TEST_FOR_KANA_MODIFIER_INSENSITIVE_CONVERSION");

    for (boolean value : new boolean[] {true, false}) {
      sharedPreferences.edit()
          .clear()
          .putBoolean(PreferenceUtil.PREF_KANA_MODIFIER_INSENSITIVE_CONVERSION_KEY, value)
          .commit();

      assertEquals(Config.newBuilder()
                       .setUseKanaModifierInsensitiveConversion(value)
                       .build(),
                   ConfigUtil.toConfig(sharedPreferences));
    }
  }

  @SmallTest
  public void testToConfig_typingCorrection() {
    SharedPreferences sharedPreferences = MozcPreferenceUtil.getSharedPreferences(
        getInstrumentation().getContext(), "TEST_FOR_TYPING_CORRECTION");

    for (boolean value : new boolean[] {true, false}) {
      sharedPreferences.edit()
          .clear()
          .putBoolean(PreferenceUtil.PREF_TYPING_CORRECTION_KEY, value)
          .commit();

      assertEquals(Config.newBuilder()
                       .setUseTypingCorrection(value)
                       .build(),
                   ConfigUtil.toConfig(sharedPreferences));
    }
  }


  @SmallTest
  public void testToConfig_historyLearningLevel() {
    SharedPreferences sharedPreferences = MozcPreferenceUtil.getSharedPreferences(
        getInstrumentation().getContext(), "TEST_FOR_HISTORY_LEARNING_LEVEL");

    for (HistoryLearningLevel historyLearningLevel : HistoryLearningLevel.values()) {
      sharedPreferences.edit()
          .clear()
          .putString(PreferenceUtil.PREF_DICTIONARY_PERSONALIZATION_KEY,
                     historyLearningLevel.name())
          .commit();

      assertEquals(Config.newBuilder()
                       .setHistoryLearningLevel(historyLearningLevel)
                       .build(),
                   ConfigUtil.toConfig(sharedPreferences));
    }
  }

  @SmallTest
  public void testToConfig_incognitoMode() {
    SharedPreferences sharedPreferences = MozcPreferenceUtil.getSharedPreferences(
        getInstrumentation().getContext(), "TEST_FOR_INCOGNITO_MODE");

    for (boolean incognitoMode : new boolean[] { true, false }) {
      sharedPreferences.edit()
          .clear()
          .putBoolean(PreferenceUtil.PREF_OTHER_INCOGNITO_MODE_KEY, incognitoMode)
          .commit();

      assertEquals(Config.newBuilder()
                       .setIncognitoMode(incognitoMode)
                       .build(),
                   ConfigUtil.toConfig(sharedPreferences));
    }
  }

  @SmallTest
  public void testToConfig_generalConfig() {
    SharedPreferences sharedPreferences = MozcPreferenceUtil.getSharedPreferences(
        getInstrumentation().getContext(), "TEST_FOR_GENERAL_CONFIG");

    for (boolean uploadUsageStats : new boolean[] { true, false }) {
      sharedPreferences.edit()
          .clear()
          .putBoolean(PreferenceUtil.PREF_OTHER_USAGE_STATS_KEY, uploadUsageStats)
          .commit();

      assertEquals(Config.newBuilder()
                       .setGeneralConfig(GeneralConfig.newBuilder()
                            .setUploadUsageStats(uploadUsageStats))
                       .build(),
                   ConfigUtil.toConfig(sharedPreferences));
    }
  }
}
